log/logging.go

```diff
@@ -211,114 +316,149 @@ func NewWithID(id string) *Logger {
+	s.logMessage(LevelWarnName, s.id, description, keysAndValues...)
 }
 
 // Info outputs an info message with an optional list of key/value pairs.
 //
 // If LogLevel is set below LevelInfo, calling this method will yield no
 // side effects.
-func (s *Logger) Info(description string, keysAndValues ...interface{}) {
-	if s.Level < LevelInfo {
+func (s *logger) Info(description string, keysAndValues ...interface{}) {
+	if s.level < LevelInfo {
 		return
 	}
-	logMessage(s.l, s.ID, "INFO ", description, keysAndValues...)
+	s.logMessage(LevelInfoName, s.id, description, keysAndValues...)
 }
 
 // Debug outputs an info message with an optional list of key/value pairs.
 //
 // If LogLevel is set below LevelDebug, calling this method will yield no
 // side effects.
-func (s *Logger) Debug(description string, keysAndValues ...interface{}) {
-	if s.Level < LevelDebug {
+func (s *logger) Debug(description string, keysAndValues ...interface{}) {
+	if s.level < LevelDebug {
 		return
 	}
-	logMessage(s.l, s.ID, "DEBUG", description, keysAndValues...)
+	s.logMessage(LevelDebugName, s.id, description, keysAndValues...)
 }
 
 // Trace outputs an info message with an optional list of key/value pairs.
 //
 // If LogLevel is set below LevelTrace, calling this method will yield no
 // side effects.
-func (s *Logger) Trace(description string, keysAndValues ...interface{}) {
-	if s.Level < LevelTrace {
+func (s *logger) Trace(description string, keysAndValues ...interface{}) {
+	if s.level < LevelTrace {
 		return
 	}
-	logMessage(s.l, s.ID, "TRACE", description, keysAndValues...)
+	s.logMessage(LevelTraceName, s.id, description, keysAndValues...)
+}
+
+func (s *logger) logMessage(level LogLevelName, id string, description string, keysAndValues ...interface{}) {
+	msg := s.formatter.formatLogEvent(s.flags, id, level, description, s.staticArgs, keysAndValues...)
+	s.l.Println(msg)
+}
+
+func (s *logger) SetLevel(level LogLevel) {
+	s.level = level
 }
 
 // SetOutput sets the output destination for the logger.
 //
 // Useful to change where the log stream ends up being written to.
-func (s *Logger) SetOutput(w io.Writer) {
-	s.l = log.New(w, defaultPrefix, s.l.Flags())
+func (s *logger) SetOutput(w io.Writer) {
+	s.l = log.New(w, s.prefix, s.flags)
 }
 
 // SetFlags changes the timestamp flags on the output of the logger.
-func (s *Logger) SetTimestampFlags(flags int) {
+func (s *logger) SetTimestampFlags(flags int) {
+	s.flags = flags
 	s.l.SetFlags(flags)
 }
 
-// logMessage writes a formatted message to the default logger.
-//
+// Add a key/value field to every log line from this logger.
+func (s *logger) SetStaticField(name string, value interface{}) {
+	s.staticArgs[name] = fmt.Sprintf("%v", value)
+}
+
+type logFormatter interface {
+	formatLogEvent(
+		flags int,
+		id string,
+		level LogLevelName,
+		description string,
+		staticFields map[string]string,
+		extraFieldKeysAndValues ...interface{},
+	) string
+}
+
+type plainTextLogFormatter struct{}
```

Comment:
Sorry to ring this bell again, but what are y'all's feels on keeping this as an interface+structs vs just use a func alias (i.e. expose that `formatLogEvent` signature as the `Formatter` itself)?

I kinda like the idea of the statelessness of formatting, deeply enforced by a function.